File: chart/js/src/spec.ts
// chart-v1 spec — the frame-side twin of chart/ssr/spec.go.
//
// The canonical doc arrives over init.doc (untrusted postMessage payload),
// so it is validated and normalized HERE before rendering. The rules are a
// deliberate mirror of the Go Normalize(): same type set, same caps
// (10,000 total points / 12 series), same defaults (tickCount 10, legend
// on), same domain rules in extents(). When the two sides drift, the
// server/client agreement test catches it — keep them in lockstep.

export const SCHEMA_VERSION = "chart-v1";
export const MAX_POINTS = 10000;
export const MAX_SERIES = 12;
export const MAX_TEXT = 200;

export type ChartType = "line" | "bar" | "area" | "scatter";

export interface Point {
  x: number;
  y: number;
}

export interface Series {
  name: string;
  points: Point[];
}

export interface AxisSpec {
  label: string;
  tickCount: number;
}

export interface Spec {
  schemaVersion: string;
  type: ChartType;
  title: string;
  series: Series[];
  axes: { x: AxisSpec; y: AxisSpec };
  options: { legend: boolean };
}

export class SpecError extends Error {}

const TYPES: Record<string, true> = { line: true, bar: true, area: true, scatter: true };

function finiteNumber(v: unknown): number | null {
  return typeof v === "number" && Number.isFinite(v) ? v : null;
}

/** Normalize an untrusted doc into a renderable Spec, or throw SpecError. */
export function normalizeSpec(doc: unknown): Spec {
  if (!doc || typeof doc !== "object") throw new SpecError("doc is not an object");
  const raw = doc as Record<string, unknown>;

  const schemaVersion = raw.schemaVersion === undefined ? SCHEMA_VERSION : String(raw.schemaVersion);
  if (schemaVersion !== SCHEMA_VERSION) {
    throw new SpecError(`unknown schemaVersion ${schemaVersion}`);
  }
  const type = String(raw.type ?? "");
  if (!TYPES[type]) throw new SpecError(`unknown chart type ${type}`);

  const title = typeof raw.title === "string" ? raw.title : "";
  if (title.length > MAX_TEXT) throw new SpecError("title too long");

  const seriesRaw = Array.isArray(raw.series) ? raw.series : null;
  if (!seriesRaw || seriesRaw.length === 0) throw new SpecError("chart has no series");
  if (seriesRaw.length > MAX_SERIES) throw new SpecError("too many series");

  const series: Series[] = [];
  let total = 0;
  for (let i = 0; i < seriesRaw.length; i++) {
    const s = seriesRaw[i];
    if (!s || typeof s !== "object") throw new SpecError(`series ${i + 1} is not an object`);
    const name = typeof (s as Record<string, unknown>).name === "string" && (s as Record<string, unknown>).name !== ""
      ? ((s as Record<string, unknown>).name as string)
      : `series ${i + 1}`;
    if (name.length > MAX_TEXT) throw new SpecError("series name too long");
    const pointsRaw = Array.isArray((s as Record<string, unknown>).points) ? ((s as Record<string, unknown>).points as unknown[]) : [];
    const points: Point[] = [];
    for (const p of pointsRaw) {
      if (!p || typeof p !== "object") throw new SpecError("point is not an object");
      const rec = p as Record<string, unknown>;
      const x = finiteNumber(rec.x);
      const y = finiteNumber(rec.y);
      if (x === null || y === null) throw new SpecError("point x/y must be finite numbers");
      points.push({ x, y });
    }
    total += points.length;
    series.push({ name, points });
  }
  if (total > MAX_POINTS) throw new SpecError("too many points");

  const axesRaw = (raw.axes && typeof raw.axes === "object" ? raw.axes : {}) as Record<string, unknown>;
  const axis = (a: unknown): AxisSpec => {
    const rec = (a && typeof a === "object" ? a : {}) as Record<string, unknown>;
    const label = typeof rec.label === "string" ? rec.label : "";
    if (label.length > MAX_TEXT) throw new SpecError("axis label too long");
    let tickCount = 10;
    const n = finiteNumber(rec.tickCount);
    if (n !== null && Number.isInteger(n)) tickCount = Math.min(20, Math.max(2, Math.trunc(n)));
    return { label, tickCount };
  };

  const optionsRaw = (raw.options && typeof raw.options === "object" ? raw.options : {}) as Record<string, unknown>;

  return {
    schemaVersion,
    type: type as ChartType,
    title,
    series,
    axes: { x: axis(axesRaw.x), y: axis(axesRaw.y) },
    options: { legend: optionsRaw.legend === undefined ? true : optionsRaw.legend !== false },
  };
}

/**
 * The domain rules — a mirror of ssr.Extents. The frame passes the result
 * to Plot as explicit `domain`, which is what pins the hydrated chart to
 * the SSR chart's extents:
 *   - bar/area include the zero baseline (Plot's maybeZero does the same)
 *   - bar pads x by half a bar group each side, so edge bars fit
 *   - a degenerate extent widens to m-1..m+1
 */
export function extents(spec: Spec): { x: [number, number]; y: [number, number] } {
  let has = false;
  let x0 = 0, x1 = 0, y0 = 0, y1 = 0;
  for (const ser of spec.series) {
    for (const p of ser.points) {
      if (!has) {
        x0 = x1 = p.x;
        y0 = y1 = p.y;
        has = true;
        continue;
      }
      x0 = Math.min(x0, p.x);
      x1 = Math.max(x1, p.x);
      y0 = Math.min(y0, p.y);
      y1 = Math.max(y1, p.y);
    }
  }
  if (!has) return { x: [0, 1], y: [0, 1] };
  if (spec.type === "bar" || spec.type === "area") {
    y0 = Math.min(y0, 0);
    y1 = Math.max(y1, 0);
  }
  if (!(x1 > x0)) { const m = x0; x0 = m - 1; x1 = m + 1; }
  if (!(y1 > y0)) { const m = y0; y0 = m - 1; y1 = m + 1; }
  if (spec.type === "bar") {
    const g = minXGap(spec);
    if (g !== Infinity) {
      x0 -= 0.4 * g;
      x1 += 0.4 * g;
    }
  }
  return { x: [x0, x1], y: [y0, y1] };
}

/**
 * Smallest positive gap between adjacent x values across all series
 * (pooled, sorted; duplicates collapse to zero and are skipped). Infinity
 * when fewer than two distinct x values exist. Mirrors ssr.minXGap — the
 * agreement test compares the padded data-domain-x attributes, so the
 * arithmetic here must produce the identical double the Go side does.
 */
export function minXGap(spec: Spec): number {
  const xs: number[] = [];
  for (const ser of spec.series) for (const p of ser.points) xs.push(p.x);
  xs.sort((a, b) => a - b);
  let minGap = Infinity;
  for (let i = 1; i < xs.length; i++) {
    const g = xs[i] - xs[i - 1];
    if (g > 0 && g < minGap) minGap = g;
  }
  return minGap;
}

File: chart/ssr/spec.go
// Package ssr renders a chart-v1 spec to a standalone SVG in pure Go.
//
// It generalizes what richtext/ssr does for documents: the same canonical
// blob the interactive frame consumes is rendered server-side, so a page
// works with JavaScript off. The renderer is pure, deterministic (same spec
// → identical bytes), and depends on nothing outside the Go standard
// library plus the framework's core/render. The plugin's two renderers —
// this package (server SVG) and the Observable Plot frame (client SVG) —
// must AGREE on axis tick labels, series count, and data extents. Agreement
// on ticks is why ticks.go is a
// faithful port of d3-array's algorithm rather than a hand-rolled
// nice-number pass; see ticks.go for the citation.
package ssr

import (
	"bytes"
	"encoding/json"
	"fmt"
	"math"
	"sort"
)

// SchemaVersion is the canonical-doc schema this renderer speaks.
const SchemaVersion = "chart-v1"

// Caps enforced at normalization. A chart's data lives in the doc (unlike
// the datagrid plugin) because a chart is small by definition — but "small"
// needs a hard edge, so normalization rejects specs past these caps.
// Both caps are part of the published contract; see docs/chart.md.
const (
	// MaxPoints is the maximum TOTAL number of points across all series.
	MaxPoints = 10000
	// MaxSeries is the maximum number of series in one chart.
	MaxSeries = 12
)

// Chart types supported by chart-v1. This is the whole set; a fifth type is
// out of scope by design (this is not a chart grammar).
const (
	TypeLine    = "line"
	TypeBar     = "bar"
	TypeArea    = "area"
	TypeScatter = "scatter"
)

// Point is one datum. Both coordinates are required and finite.
type Point struct {
	X *float64 `json:"x"`
	Y *float64 `json:"y"`
}

// Series is a named sequence of points.
type Series struct {
	Name   string  `json:"name"`
	Points []Point `json:"points"`
}

// Axis carries an optional caption and the approximate tick count. The tick
// count is passed VERBATIM to the tick algorithm on both renderers, which is
// what keeps server and client labels identical.
type Axis struct {
	Label     string `json:"label"`
	TickCount *int   `json:"tickCount"`
}

// Axes describes both axes.
type Axes struct {
	X Axis `json:"x"`
	Y Axis `json:"y"`
}

// Options carries rendering hints. Width/Height are the SVG viewBox size
// (the SVG itself is width="100%" and scales).
type Options struct {
	Legend *bool    `json:"legend,omitempty"`
	Width  *float64 `json:"width,omitempty"`
	Height *float64 `json:"height,omitempty"`
}

// Spec is the canonical chart document.
type Spec struct {
	SchemaVersion string   `json:"schemaVersion"`
	Type          string   `json:"type"`
	Title         string   `json:"title"`
	Series        []Series `json:"series"`
	Axes          Axes     `json:"axes"`
	Options       Options  `json:"options"`
}

// ParseSpec decodes and normalizes a chart-v1 spec from JSON. Every rule
// here has a twin in the frame bundle (chart/js/src/spec.ts); the two MUST
// stay in lockstep because agreement between the renderers is the product.
func ParseSpec(raw []byte) (Spec, error) {
	var s Spec
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	if err := dec.Decode(&s); err != nil {
		return Spec{}, fmt.Errorf("ssr: invalid spec JSON: %w", err)
	}
	return Normalize(s)
}

// Normalize validates and fills defaults. It is pure: no mutation of the
// input's backing arrays.
func Normalize(s Spec) (Spec, error) {
	if s.SchemaVersion == "" {
		s.SchemaVersion = SchemaVersion
	}
	if s.SchemaVersion != SchemaVersion {
		return Spec{}, fmt.Errorf("ssr: unknown schemaVersion %q (want %q)", s.SchemaVersion, SchemaVersion)
	}
	switch s.Type {
	case TypeLine, TypeBar, TypeArea, TypeScatter:
	default:
		return Spec{}, fmt.Errorf("ssr: unknown chart type %q (want line|bar|area|scatter)", s.Type)
	}
	if len(s.Series) == 0 {
		return Spec{}, fmt.Errorf("ssr: chart has no series")
	}
	if len(s.Series) > MaxSeries {
		return Spec{}, fmt.Errorf("ssr: %d series exceeds the %d-series cap", len(s.Series), MaxSeries)
	}
	total := 0
	for i := range s.Series {
		if s.Series[i].Name == "" {
			s.Series[i].Name = fmt.Sprintf("series %d", i+1)
		}
		if len(s.Series[i].Name) > maxTextRunes {
			return Spec{}, fmt.Errorf("ssr: series %d name exceeds %d characters", i+1, maxTextRunes)
		}
		for j, p := range s.Series[i].Points {
			if p.X == nil || p.Y == nil {
				return Spec{}, fmt.Errorf("ssr: series %q point %d: x and y are both required", s.Series[i].Name, j)
			}
			if math.IsNaN(*p.X) || math.IsInf(*p.X, 0) || math.IsNaN(*p.Y) || math.IsInf(*p.Y, 0) {
				return Spec{}, fmt.Errorf("ssr: series %q point %d: coordinates must be finite", s.Series[i].Name, j)
			}
		}
		total += len(s.Series[i].Points)
	}
	if total > MaxPoints {
		return Spec{}, fmt.Errorf("ssr: %d total points exceeds the %d-point cap", total, MaxPoints)
	}
	if n := len(s.Title); n > maxTextRunes {
		return Spec{}, fmt.Errorf("ssr: title exceeds %d characters", maxTextRunes)
	}
	if n := len(s.Axes.X.Label); n > maxTextRunes {
		return Spec{}, fmt.Errorf("ssr: x-axis label exceeds %d characters", maxTextRunes)
	}
	if n := len(s.Axes.Y.Label); n > maxTextRunes {
		return Spec{}, fmt.Errorf("ssr: y-axis label exceeds %d characters", maxTextRunes)
	}
	s.Axes.X.TickCount = clampTickCount(s.Axes.X.TickCount)
	s.Axes.Y.TickCount = clampTickCount(s.Axes.Y.TickCount)
	if s.Options.Legend == nil {
		def := true
		s.Options.Legend = &def
	}
	if s.Options.Width == nil {
		w := defaultWidth
		s.Options.Width = &w
	} else if *s.Options.Width < minWidth || *s.Options.Width > maxWidth {
		return Spec{}, fmt.Errorf("ssr: width %g outside [%g, %g]", *s.Options.Width, minWidth, maxWidth)
	}
	if s.Options.Height == nil {
		h := defaultHeight
		s.Options.Height = &h
	} else if *s.Options.Height < minHeight || *s.Options.Height > maxHeight {
		return Spec{}, fmt.Errorf("ssr: height %g outside [%g, %g]", *s.Options.Height, minHeight, maxHeight)
	}
	return s, nil
}

// maxTextRunes caps user text (title, axis labels, series names). Go's len
// counts bytes; cap on runes so multi-byte names get a fair allowance.
const maxTextRunes = 200

const (
	defaultWidth   = 720.0
	defaultHeight  = 420.0
	minWidth       = 200.0
	maxWidth       = 1200.0
	minHeight      = 120.0
	maxHeight      = 900.0
	defaultTicks   = 10
	minTicks, maxT = 2, 20
)

func clampTickCount(n *int) *int {
	if n == nil {
		v := defaultTicks
		return &v
	}
	if *n < minTicks {
		v := minTicks
		return &v
	}
	if *n > maxT {
		v := maxT
		return &v
	}
	return n
}

// Extents returns the x and y domains the renderers use. Both renderers
// compute domains from these rules (and the frame is given the domain
// explicitly), so they agree by construction:
//
//   - x: min..max of x across all series
//   - y: min..max of y across all series
//   - bar/area include the zero baseline (Plot's maybeZero does the same)
//   - bar pads x by half a bar group on each side, so the first and last
//     bars sit fully inside the plot instead of straddling the axis
//   - a degenerate extent widens to m-1..m+1
func Extents(s Spec) (x, y [2]float64) {
	has := false
	var x0, x1, y0, y1 float64
	for _, ser := range s.Series {
		for _, p := range ser.Points {
			if !has {
				x0, x1, y0, y1 = *p.X, *p.X, *p.Y, *p.Y
				has = true
				continue
			}
			x0 = math.Min(x0, *p.X)
			x1 = math.Max(x1, *p.X)
			y0 = math.Min(y0, *p.Y)
			y1 = math.Max(y1, *p.Y)
		}
	}
	if !has {
		return [2]float64{0, 1}, [2]float64{0, 1}
	}
	if s.Type == TypeBar || s.Type == TypeArea {
		y0 = math.Min(y0, 0)
		y1 = math.Max(y1, 0)
	}
	if x1 <= x0 {
		x0, x1 = x0-1, x0+1
	}
	if y1 <= y0 {
		y0, y1 = y0-1, y0+1
	}
	if s.Type == TypeBar {
		if g := minXGap(s); !math.IsInf(g, 1) {
			x0 -= 0.4 * g
			x1 += 0.4 * g
		}
	}
	x = [2]float64{x0, x1}
	y = [2]float64{y0, y1}
	return x, y
}

// minXGap returns the smallest positive gap between adjacent x values
// across all series (x values pooled and sorted; duplicates collapse to a
// zero gap and are skipped). +Inf when fewer than two distinct x values
// exist. It sizes bar groups and the bar-type x padding; the frame bundle's
// twin (chart/js/src/spec.ts minXGap) computes the identical double — the
// agreement test compares the padded data-domain-x attributes, so the two
// implementations must stay in lockstep.
func minXGap(s Spec) float64 {
	var xs []float64
	for _, ser := range s.Series {
		for _, p := range ser.Points {
			xs = append(xs, *p.X)
		}
	}
	sort.Float64s(xs)
	minGap := math.Inf(1)
	for i := 1; i < len(xs); i++ {
		if g := xs[i] - xs[i-1]; g > 0 && g < minGap {
			minGap = g
		}
	}
	return minGap
}
